torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp

```diff
@@ -165,97 +205,262 @@ bool frozenMkldnnCompatibleConvNode(Node* n) {
+  }
 
-    if (frozenMkldnnCompatibleLinearNode(n)) {
-      convertFrozenLinearToMKLDNN(n);
+  std::pair<graph_node_list::iterator, bool> scanNode(Node* producer) {
+    if (MKLDNNGroupStart(producer)) {
+      if (producer->kind() != prim::MKLDNNGroup) {
+        producer = SubgraphUtils::createSingletonSubgraphAndUpdateAliasing(
+            producer, prim::MKLDNNGroup, aliasDb_);
+      }
+      std::vector<Node*> output_nodes;
+      for (Value* v : producer->outputs()) {
+        for (const auto& use : v->uses()) {
+          output_nodes.push_back(use.user);
+        }
+      }
+      std::sort(
+          output_nodes.begin(), output_nodes.end(), [&](Node* a, Node* b) {
+            return a->isBefore(b);
+          });
+      for (auto output_node : output_nodes) {
+        if (auto group = tryMerge(producer, output_node)) {
+          // we successfully merged, so the new group's `outputs` may have
+          // changed. So rescan the new group for more merging opportunities.
+          return std::make_pair(group.value()->iterator()++, true);
+        }
+      }
     }
+
+    return std::make_pair(++producer->iterator(), false);
   }
-}
 
-void ConvertFrozenOpsToMKLDNN(Block* b) {
+  // Try to merge `consumer` into `producer`. If successful, this destroys
+  // `consumer` and returns the `producer` group.
+  c10::optional<Node*> tryMerge(Node* producer, Node* consumer) {
+    AT_ASSERT(producer->kind() == prim::MKLDNNGroup);
+    bool canMerge = shouldConsiderForMerge(consumer) &&
+        aliasDb_.moveAfterTopologicallyValid(consumer, producer);
+
+    if (!canMerge) {
+      return c10::nullopt;
+    }
+
+    // TODO: subgraph utils doesnt handle merging a consumer into a producer
+    // so we merge the producer into the consumer even though it's inefficient
+    consumer = SubgraphUtils::createSingletonSubgraphAndUpdateAliasing(
+        consumer, prim::MKLDNNGroup, aliasDb_);
+
+    SubgraphUtils::mergeNodeIntoSubgraphAndUpdateAliasing(
+        producer, consumer, aliasDb_);
+    return consumer;
+  }
+
+  Block* block_;
+  std::shared_ptr<Graph> graph_;
+  AliasDb& aliasDb_;
+};
+
+bool containsMKLDNNGroup(Block* b) {
   for (Node* n : b->nodes()) {
     for (Block* block : n->blocks()) {
-      ConvertFrozenOpsToMKLDNN(block);
+      if (containsMKLDNNGroup(block)) {
+        return true;
+      }
     }
-
-    if (frozenMkldnnCompatibleConvNode(n)) {
-      convertFrozenConvToMKLDNN(n);
+    if (MKLDNNSubgraphSlicer::MKLDNNGroupStart(n)) {
+      return true;
     }
   }
+  return false;
 }
 
 } // namespace
 
 void ConvertFrozenOpsToMKLDNN(std::shared_ptr<Graph>& graph) {
 #if AT_MKLDNN_ENABLED()
+  GRAPH_DUMP("Before convert frozen ops to mkldnn", graph);
+  // TODO: replace conv1d with conv2d ?
   graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
-  ConvertFrozenConvParamsToMKLDNN(graph->block());
-  EliminateDeadCode(graph);
+  if (containsMKLDNNGroup(graph->block())) {
```

Comment:
I wonder if it might also make sense to have another check similar `tensorInputIsMKLDNNSupported` to make a bit more conservative?

Reply:
That check is covered as part of this call